### PR TITLE
Add new lmdb tables for work order processing

### PR DESCRIPTION
**Pull request by rranjan3**
- Add MDB_NOTFOUND check in db_store_get_value_size and return Success
       if databse itself not found.
- Log instead of suppressing exception at shared_kv_db_store
- Toggle sync mode to use polling
- Add tables wo-worker-pending and wo-worker-processing
- Update usage of these tables in in enclave_manager
- Update generic client to pick up workers randomly

### CODE REVIEW

**Comment by rranjan3** [comment]
Cleanup at startup of enclaves and work-order handler needs to be revisited to remove the usage of wo_scheduled table.

File: enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py [hunk 1]
```diff
@@ -82,53 +82,43 @@ def _create_signup_data(self):
...
+        while wo_id is not None:
+            if process_wo_id == wo_id:
+                self._kv_helper.set("wo-worker-processing",
+                                    self._worker_id, wo_id)
+                return self._process_work_order_by_id(wo_id)
             else:
                 return None
-        # end of for loop
+            wo_id = self._kv_helper.csv_pop(
+                "wo-worker-pending", self._worker_id)
```
[the remark this answers is not in the document]
Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py:103]: Yes @manojgop csv_pop would remove from the wo-worker-pending table. But with singleton, the if check is more of a double check only as there is no contending worker to pick up the same work-order.
However, even with WPE, this behavior is same as of now as sync mode is being restricted to pools with a single WPE. When there are multiple WPE contending over the same queue, this will be an issue.

File: enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py [hunk 2]
```diff
@@ -152,23 +142,25 @@ def _process_work_order_by_id(self, wo_id):
                 logger.error("Received empty work order corresponding " +
                              "to id %s from wo-requests table", wo_id)
                 self._kv_helper.remove("wo-processing", wo_id)
+                # Remove from wo-worker-processing
+                self._kv_helper.remove("wo-worker-processing", self._worker_id)
```
[the remark this answers is not in the document]
Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py:149]: Yes both are required though renaming one would help. I plan to rename` wo-processing` to `wo-in-process-status` or `wo-status`. It holds the Enum `WorkOrderStatus` for work orders that are at some stage of processing.

[the remark this answers is not in the document]
Comment (manojgop) [reply enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py:149]: You could keep same name for old table. May be new table can be renamed to wo-worker-map

[the remark this answers is not in the document]
Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py:149]: In that case I will update straightaway.

File: enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (manojgop) [reply enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py:103]: We might have multiple singleton workers and a combination of singleton and KME/WPE. I would suggest to have an API which can directly fetch `process_wo_id` from the lmdb table for sync mode. lmdb access can be a remote call over the network  and it's costly to loop over all the work orders by making multiple network calls.

[the remark this answers is not in the document]
Comment (manojgop) [reply enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py:103]: Also for WPE, it's more efficient to fetch a work order from Lmdb which this WPE  can support . For example, WPE can have a config parameter of list of workload_ids which it can support and fetch a work order id which matches the workload id .

File: enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py [hunk 3]
```diff
@@ -217,6 +216,8 @@ def _process_work_order_by_id(self, wo_id):
         logger.info("Delete workorder entry %s from wo-processing table",
                     wo_id)
         self._kv_helper.remove("wo-processing", wo_id)
+        # Remove from wo-worker-map
+        self._kv_helper.remove("wo-worker-map", self._worker_id)
```
Comment (manojgop) [inline enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py:220]: `wo-worker-map` is not really getting used here. Entries are added and removed from this table but not read anywhere.  So can we remove it from this PR ?

File: enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py [hunk 4]
```diff
@@ -94,53 +99,48 @@ def _manager_on_boot(self):
...
-            logger.error("Problem while getting ids from wo-scheduled table")
-            return
-
-        for wo_id in list_of_workorders:
-            if wo_id == process_wo_id:
-                resp = self._process_work_order_by_id(wo_id)
-                return resp
+        logger.info("Processing work orders found in wo-worker-pending table.")
+
+        wo_id = self._kv_helper.csv_pop("wo-worker-pending", self._worker_id)
```
Comment (manojgop) [inline enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py:109]: can we use `csv_peek` and `csv_pop` if  `process_wo_id == wo_id`

File: enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py [hunk 5]
```diff
@@ -94,53 +99,48 @@ def _manager_on_boot(self):
...
-        except Exception as e:
-            logger.error("Problem while getting keys from wo-scheduled table")
-            return
+        logger.info("Processing work orders found in wo-worker-pending table.")
 
-        for wo_id in list_of_workorders:
+        wo_id = self._kv_helper.csv_pop("wo-worker-pending", self._worker_id)
+        while wo_id is not None:
+            self._kv_helper.set("wo-worker-map",
+                                self._enclave_id_hex, wo_id)
```
Comment (manojgop) [inline enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py:138]: Can we use self._worker_id instead of self._enclave_id_hex to be consistent with singleton

Comment (manojgop) [reply enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py:138]: You could remove `wo-worker-map` from this PR as it's not getting used

File: shared_kv_storage/db_store/packages/lmdb_store.cpp [hunk 6]
```diff
@@ -168,7 +168,12 @@ tcf_err_t db_store::db_store_get_value_size(
     }
 
     ret = mdb_dbi_open(stxn.txn, table.c_str(), 0, &dbi);
-    if (ret != 0) {
+    if (ret == MDB_NOTFOUND) {
+        // Return Success if database not found
```
Comment (manojgop) [inline shared_kv_storage/db_store/packages/lmdb_store.cpp:172]: Not sure why are we returning success if database/table is not found

File: enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py [hunk 7]
```diff
@@ -164,23 +164,26 @@ def _process_work_order_by_id(self, wo_id):
...
         except Exception as e:
             logger.error("Problem while reading the work order %s "
                          "from wo-requests table", wo_id)
             self._kv_helper.remove("wo-processing", wo_id)
+            # Remove from wo-worker-map
+            # @TODO : Add to wo-worker-pending if the error occurred is
+            # recoverable, i.e.-retrying again could succeed.
+            self._kv_helper.remove("wo-worker-map",
+                                   self._enclave_id_hex)
             return
```
Comment (Ram-srini) [inline enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py:180]: Return value of _process_work_order_by_id() is inconsistent. In case of error it is returning without value and on success it is returning response. And caller side we are returning as it is.

File: shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py [hunk 8]
```diff
@@ -82,7 +82,9 @@ def set(self, table, key, value):
         try:
             self._db_store.db_store_put(table, key, value)
             return True
-        except Exception:
+        except Exception as ex:
+            logger.warn("Exception occured in set - {}".
```
Comment (Ram-srini) [inline shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py:86]: is the log level function correct in case of exception?

File: enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py (hunk 3 above)
Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py]: Removed.

File: enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py (hunk 4 above)
Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py]: Using csv_match_pop which does both this in a single remote db call.

File: enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py (hunk 5 above)
Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py]: > Can we use self._worker_id instead of self._enclave_id_hex to be consistent with singleton

_worker_id is explicitly needed by WPE as well.

File: enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py [hunk 9]
```diff
@@ -82,53 +82,46 @@ def _create_signup_data(self):
...
-            logger.error("Problem while getting ids from wo-scheduled table")
-            return
-
-        for wo_id in list_of_workorders:
-            if wo_id == process_wo_id:
-                resp = self._process_work_order_by_id(wo_id)
-                return resp
+        logger.info("Processing work orders found in wo-worker-pending table.")
+
+        wo_id = self._kv_helper.csv_peek("wo-worker-pending", self._worker_id)
```
[the remark this answers is not in the document]
Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py]: Using csv_match_pop() instead.. that does both in one remote db call. It matches and returns value if match is successful. Implementation/PR is work in progress for csv_match_pop() API.

File: config/singleton_enclave_config.toml [hunk 10]
```diff
@@ -28,7 +28,7 @@ LogFile  = "__screen__"
 # ------------------------------------------------------------------
 [WorkloadExecution]
 # Set this flag to 1 for sync workorder execution.Else set it to 0
-sync_workload_execution = 1
+sync_workload_execution = 0
```
[the remark this answers is not in the document]
Comment (rranjan3) [reply config/singleton_enclave_config.toml:31]: I have disabled it for the time being @Ram-srini for work in progress https://github.com/hyperledger/avalon/pull/536#discussion_r441518460.

File: enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py [hunk 11]
```diff
@@ -82,53 +82,46 @@ def _create_signup_data(self):
     def _process_work_order_sync(self, process_wo_id):
         """
         Process the work-order of the specified work-order id
-        and return the response.
-        Used for synchronous execution.
+        and return the response. Used for synchronous execution.
```
[the remark this answers is not in the document]
Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py:85]: Verified polling mode after updating to new API. Before introduction of new API, sync mode was working fine.

File: shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py (hunk 8 above)
Comment (rranjan3) [reply shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py:86]: Do you recommend error? We ignore it anyways. I introduced the logs just for knowledge as the flow would be fine as return is taking care of it.

File: enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py [hunk 12]
```diff
@@ -94,53 +99,43 @@ def _manager_on_boot(self):
...
-                logger.info("Received empty list of work orders from " +
-                            "wo-scheduled table")
-                return
-        except Exception as e:
-            logger.error("Problem while getting keys from wo-scheduled table")
-            return
+        logger.info("Processing work orders found in wo-worker-pending table.")
 
-        for wo_id in list_of_workorders:
+        wo_id = self._kv_helper.csv_pop("wo-worker-pending", self._worker_id)
```
Comment (manojgop) [inline enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py:132]: WPE needs to check if it can support workload id present in the work order. Please add a TODO comment if you plan to do this in different PR

File: enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py [hunk 13]
```diff
@@ -42,6 +43,10 @@ class WorkOrderProcessorEnclaveManager(EnclaveManager):
 
     def __init__(self, config):
         super(WorkOrderProcessorEnclaveManager, self).__init__(config)
+        # Calculate sha256 of enclave id to get 32 bytes. Then take a
+        # hexdigest for hex str.
+        enclave_id_utf = self.enclave_id.encode("UTF-8")
+        self._enclave_id_hex = hashlib.sha256(enclave_id_utf).hexdigest()
```
Comment (manojgop) [inline enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py:49]: Are we using `_enclave_id_hex ` this in this PR ? If not, can we remove it?

Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py]: Not using. Removed.

File: enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py (hunk 12 above)
Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/wpe/wpe_enclave_manager.py:126]: Updated.

File: enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py [hunk 14]
```diff
@@ -82,53 +82,43 @@ def _create_signup_data(self):
...
-                resp = self._process_work_order_by_id(wo_id)
-                return resp
+        logger.info("Processing work orders found in wo-worker-pending table.")
+
+        # csv_match_pop will peek into the deque and pop out the first
+        # work-order id if it matches the process_wo_id
+        wo_id = self._kv_helper.csv_match_pop(
+            "wo-worker-pending", self._worker_id, process_wo_id)
+        while wo_id is not None:
+            if process_wo_id == wo_id:
```
[the remark this answers is not in the document]
Comment (rranjan3) [reply enclave_manager/avalon_enclave_manager/singleton/singleton_enclave_manager.py]: Yes redundant indeed.

File: shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py [hunk 15]
```diff
@@ -188,7 +198,9 @@ def csv_prepend(self, table, key, value):
         try:
             self._db_store.db_store_csv_prepend(table, key, value)
             return True
-        except Exception:
+        except Exception as ex:
+            logger.warn("Exception occured in csv_prepend - {}".
```
Comment (danintel) [inline shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py:202]: s/occured/occurred/

File: shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py [hunk 16]
```diff
@@ -213,7 +225,9 @@ def csv_pop(self, table, key):
                 value = self._db_store.db_store_csv_pop(table, key)
             else:
                 value = None
-        except Exception:
+        except Exception as ex:
+            logger.warn("Exception occured in csv_pop - {}".
```
Comment (danintel) [inline shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py:229]: s/occured/occurred/

File: shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py (hunk 8 above)
Comment (danintel) [reply shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py:86]: s/occured/occurred/

File: shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py [hunk 17]
```diff
@@ -99,7 +101,9 @@ def get(self, table, key):
                 value = self._db_store.db_store_get(table, key)
             else:
                 value = None
-        except Exception:
+        except Exception as ex:
+            logger.warn("Exception occured in get - {}".
```
Comment (danintel) [inline shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py:105]: s/occured/occurred/

File: shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py [hunk 18]
```diff
@@ -128,7 +132,9 @@ def remove(self, table, key, value=None):
             else:
                 self._db_store.db_store_del(table, key, value)
             return True
-        except Exception:
+        except Exception as ex:
+            logger.warn("Exception occured in remove - {}".
```
Comment (danintel) [inline shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py:136]: s/occured/occurred/

File: shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py [hunk 19]
```diff
@@ -146,7 +152,9 @@ def lookup(self, table):
             table_keys = table_keys.split(",")
             for i in table_keys:
                 result.append(i)
-        except Exception:
+        except Exception as ex:
+            logger.warn("Exception occured in lookup - {}".
```
Comment (danintel) [inline shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py:156]: s/occured/occurred/

Comment (rranjan3) [reply shared_kv_storage/kv_storage/remote_lmdb/shared_kv_dbstore.py]: Thanks @danintel. Updated.
